tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -3310,6 +3310,167 @@ llvm::Optional<Value> convertFakeQuantOp(PatternRewriter& rewriter,
       .getResult();
 }
 
+llvm::Optional<Value> convertMirrorPadCommon(PatternRewriter& rewriter,
+                                             Operation* op,
+                                             RankedTensorType output_type,
+                                             Value input, Value pad,
+                                             StringRef mode) {
+  RankedTensorType input_type = input.getType().dyn_cast<RankedTensorType>();
+  if (!input_type) {
+    (void)rewriter.notifyMatchFailure(op, "input type isn't a ranked tensor");
+    return llvm::None;
+  }
+
+  if ((mode != "SYMMETRIC") && (mode != "REFLECT")) {
+    (void)rewriter.notifyMatchFailure(op,
+                                      "mode isn't one of REFLECT or SYMMETRIC");
+    return llvm::None;
+  }
+
+  ElementsAttr pad_elems;
+  if (!matchPattern(pad, m_Constant(&pad_elems))) {
+    (void)rewriter.notifyMatchFailure(op, "fail to retrieve padding values");
+    return llvm::None;
+  }
+
+  // Indicate the value and shape of the tensor being processed in that
+  // dimension
+  Value current_tensor(input);
+  SmallVector<int64_t> current_dim_size;
+  for (int axis = 0; axis < input_type.getRank(); axis++) {
+    current_dim_size.push_back(input_type.getShape()[axis]);
+  }
+
+  llvm::Optional<Value> result;
+  const int offset = (mode == "SYMMETRIC") ? 0 : 1;
+
+  // Split N-rank mirrorpad into N sequences of "slice, reverse, concat"
+  // operation. Take an example below. Perform mirrorpad on a 2D input tensor
+  // with paddings[] = ([[1, 1,], [2, 2]]) in `SYMMETRIC` mode. Original input
+  // tensor:
+  //            | 1 2 3 |
+  //            | 4 5 6 |
+  //
+  // First, fill in the padding value on the top and bottom.
+  // pad before | 1 2 3 |
+  //            | 1 2 3 |
+  //            | 4 5 6 |
+  // pad after  | 4 5 6 |
+  //
+  // Second, fill in the padding value on the left and right.
+  // pad before          pad after
+  //       2 1 | 1 2 3 | 3 2
+  //       2 1 | 1 2 3 | 3 2
+  //       5 4 | 4 5 6 | 6 5
+  //       5 4 | 4 5 6 | 6 5
+
+  SmallVector<int64_t, 5> slice_before_begin, slice_before_size;
+  SmallVector<int64_t, 5> slice_after_begin, slice_after_size;
+  for (int axis = 0; axis < input_type.getRank(); axis++) {
```

Comment:
Avoid checking end except where it can change during loop iteration (https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop). Nit: I prefer pre increment unless post behavior is needed.

Reply:
That is a great habit. In complex cases, the loads might not be optimized out due to compilers failing to recognize the loop invariant terminating condition.